src/sumo_gym/utils/xml_utils.py

```diff
@@ -19,19 +20,78 @@ def encode_xml_fmp(net_xml_file_path: str = None, flow_xml_file_path: str = None
     # TODO
     pass
 
-def decode_xml_fmp(net_xml_file_path: str = None, flow_xml_file_path: str = None):
-    # TODO
-    # decode xml into FMP problem setting, for specific types, refer to typing.py and VRP decoding below
-    pass
+
+
+def get_edges_and_lanes(net_xml_tree):
+    """
+    Given next_xml_tree, 
+    return dictionary of edges and lanes
+
+    Assumes no edges contain stopOffset elements
+    """
+    # map edge id to dictionary with attributes
+    edges_and_lanes = {}
+
+    edge_attr = ["from", "to", "priority", "function"]
+    lane_attr = ["index", "speed", "length", "shape"]
+
+    edges = net_xml_tree.findall("edge")
+    print(edges)
+    for edge in edges:
+
+        edge_id = edge.attrib["id"]
+        val = edge.attrib
+        del val["id"]
+        
+        if "priority" in val:
+            val["priority"] = int(val["priority"])
+        
+        edges_and_lanes[edge_id] = val
+
+        lanes = edge.findall("lane")
+
+        for lane in lanes:
+
+            val = lane.attrib
+            
+            if "index" in val:
+                val["index"] = int(val["index"])
+            
+            for float_attr in ["speed", "length"]:
+                if float_attr in val:
+                    val[float_attr] = float(val[float_attr])
+
+            lane_list = edges_and_lanes[edge_id].get("lanes", [])
+            lane_list.append(val)
+            edges_and_lanes[edge_id]["lanes"] = lane_list
```

Comment:
you can check the variables and variable typing we defined in fmp.py and typing.py, basically this function should return the variable we want. Some of them in sumo is irrelevant to our FMP problem. For example, regarding edges_and_lanes, right now we only need the vertices for each edge. And you need to extract the vertices position (x, y location) from net.xml under asset/data. Most of the part is similar to what I did in original decode_xml function. But some typing is different, and there might be future changes, but some of them you should be able to retrieve for sure. Like:
- vertices
- edges
- demand at each node (you need to check how to add this to the sumo net.xml, like what I did currently, under <param> tag, but for fmp, we need it to be a tuple, with start and end. So list of <param> for each node under sumo xml, each as a tuple[start, end])
- vehicle
- charging level for vehicle (similar to above, you need to find a way to add this in sumo, as far as I know, you are able to create <param> tag for demand in Netedit! Lucky!)

Notice that we might not need capacity there, since in FMP, all capacity for vehicle is 1. 

You might find some typing or return variables not expected, feel free to change it!